include/graphblas/algorithms/triangle_count.hpp

```diff
@@ -0,0 +1,249 @@
+			const Matrix< D2, config::default_backend, RIT2, CIT2, NIT2 > & MXM_lhs,
+			const Matrix< D3, config::default_backend, RIT3, CIT3, NIT3 > & MXM_rhs,
+			Matrix< D4, config::default_backend, RIT4, CIT4, NIT4 > & EWA_out,
+			const Matrix< D5, config::default_backend, RIT5, CIT5, NIT5 > & EWA_rhs,
+			const D6 div_factor,
+			const Semiring mxm_semiring = Semiring(),
+			const MulMonoid ewiseapply_monoid = MulMonoid(),
+			const SumMonoid sumreduce_monoid = SumMonoid()
+		) {
+			if( ( &MXM_out == &MXM_lhs ) || ( &MXM_out == &MXM_rhs ) ) {
+				return ILLEGAL;
+			}
+
+			RC rc = SUCCESS;
+
+			// Compute MXM_out = Mlhs * Mrhs
+			rc = rc ? rc : mxm< descr_mxm >( MXM_out, MXM_lhs, MXM_rhs, mxm_semiring, Phase::RESIZE );
+			rc = rc ? rc : mxm< descr_mxm >( MXM_out, MXM_lhs, MXM_rhs, mxm_semiring, Phase::EXECUTE );
+
+			// Compute MXM_out .*= EWA_rhs
+			// FIXME: Replace by a foldl( Matrix[in,out], Matrix[in], Monoid ) - not implemented yet
+			// Will then become:
+			// rc = rc ? rc : eWiseApply< descr_ewa >( MXM_out, MXM_out, EWA_rhs, ewiseapply_monoid, Phase::RESIZE );
+			// rc = rc ? rc : eWiseApply< descr_ewa >( MXM_out, MXM_out, EWA_rhs, ewiseapply_monoid, Phase::EXECUTE );
+			// Instead of:
+			rc = rc ? rc : eWiseApply< descr_ewa >( EWA_out, MXM_out, EWA_rhs, ewiseapply_monoid, Phase::RESIZE );
+			rc = rc ? rc : eWiseApply< descr_ewa >( EWA_out, MXM_out, EWA_rhs, ewiseapply_monoid, Phase::EXECUTE );
+
+			// Compute a sum reduction over <EWA_out> into <count>
+			count = static_cast< size_t >( 0 );
+			rc = rc ? rc : foldl< descr_reduce >( count, EWA_out, sumreduce_monoid );
+
+			// Apply the div_factor to the reduction result
+			count /= div_factor;
+
+			return rc;
+		}
+
+		/**
+		 * Given a graph, indicates how many triangles are contained within.
+		 *
+		 * @tparam D 				The type of the matrix non-zero values.
+		 *
+		 * @param[out]    count     The number of triangles.
+		 * 						    Any prior contents will be ignored.
+		 * @param[in]     A         The input graph.
+		 * @param[in,out] MXM_out    Buffer matrix with the same dimensions as the input
+		 * 							graph. Any prior contents will be ignored.
+		 * @param[in] L 		Lower triangular matrix of the input graph (optional)
+		 * @param[in] U 		Lower triangular matrix of the input graph (optional)
+		 *
+		 *
+		 * @returns #grb::SUCCESS  When the computation completes successfully.
+		 * @returns #grb::MISMATCH If the dimensions of the input matrices/buffers
+		 * 						   are incompatible.
+		 * @returns #grb::ILLEGAL  If the given algorithm does not exist.
+		 * @returns #grb::PANIC    If an unrecoverable error has been encountered. The
+		 *                         output as well as the state of ALP/GraphBLAS is
+		 *                         undefined.
+		 *
+		 * \par Performance semantics
+		 *
+		 *   -# This function does not allocate nor free dynamic memory, nor shall it
+		 *      make any system calls.
+		 *
```

Comment:
There is a resize phase which can actually dynamically allocate memory. Therefore, the #grb::OUT_OF_MEMORY error code should be added to the directly preceding list of return options, where the highlighted text here should mention that the performance here strongly depends on the sparsity structure, which affects all of work, data movement, and memory usage costs. The only thing that is kept constant in all algorithms is the number of synchronisations needed in case there are more than 1 user processes-- namely, those required for one mxm, one eWiseApply (or foldl), and one reduction. ("For BSP1D, this amounts to 2 synchronisations." could be added as a `\note`)